### PR TITLE
sock_async: add optional callback argument

### PR DESCRIPTION
**Pull request by miri64**
### Contribution description
In https://github.com/RIOT-OS/RIOT/pull/12907/files#r390451905 I noticed that there is simply no way to get the surrounding DTLS sock from within the UDP sock's asynchronous callback. As such provides this change a way to provide context argument with the callback, similar as we already do it for `xtimer` or other asynchronous operations (and e.g. `lwip` does to for the asynchronous component of its `sock` equivalent `netconn`).

As `sock_async` is still experimental this API change should only require one ACK.

### Testing procedure
The following applications should still compile for all supported boards

- `examples/gcoap`
- `tests/gnrc_sock_async_event`
- `tests/lwip`

I amended the tests for GNRC's and lwIP's `sock_async` implementations if the callback argument supplied is correctly handed to the callback. As such the following tests should still pass

- `tests/gnrc_sock_async_event`
- `tests/lwip`

(`make -C tests/lwip tests` only succeeds on `native` and requires to have `dist/tools/tapsetup/tapsetup` run first)

### Issues/PRs references
Requirement for this PR came up in https://github.com/RIOT-OS/RIOT/pull/12907/files#r390451905. As such, #12907 depends on this PR.

### CODE REVIEW

File: pkg/lwip/contrib/sock/tcp/lwip_sock_tcp.c [hunk 1]
```diff
@@ -380,13 +380,16 @@ ssize_t sock_tcp_write(sock_tcp_t *sock, const void *data, size_t len)
 }
 
 #ifdef SOCK_HAS_ASYNC
-void sock_tcp_set_cb(sock_tcp_t *sock, sock_tcp_cb_t cb)
+void sock_tcp_set_cb(sock_tcp_t *sock, sock_tcp_cb_t cb, void *arg)
```
Comment (benpicco) [inline pkg/lwip/contrib/sock/tcp/lwip_sock_tcp.c]: Why do you provide `arg` to `sock_tcp_set_cb()` when it's not used?

File: pkg/lwip/contrib/sock/tcp/lwip_sock_tcp.c [hunk 2]
```diff
@@ -380,13 +380,16 @@ ssize_t sock_tcp_write(sock_tcp_t *sock, const void *data, size_t len)
...
 {
+    (void)arg;
     sock->base.async_cb.tcp = cb;
 }
 
-void sock_tcp_queue_set_cb(sock_tcp_queue_t *queue, sock_tcp_queue_cb_t cb)
+void sock_tcp_queue_set_cb(sock_tcp_queue_t *queue, sock_tcp_queue_cb_t cb,
+                           void *arg)
```
Comment (benpicco) [inline pkg/lwip/contrib/sock/tcp/lwip_sock_tcp.c]: Why does `sock_tcp_set_cb()` only set the callback, but `sock_tcp_queue_set_cb()` sets callback and arg?

File: pkg/lwip/contrib/sock/tcp/lwip_sock_tcp.c (hunk 1 above)
Comment (miri64) [reply pkg/lwip/contrib/sock/tcp/lwip_sock_tcp.c]: Oops, this is an error from when I progressively introduced this in my commit history.

Comment (miri64) [reply pkg/lwip/contrib/sock/tcp/lwip_sock_tcp.c]: Interesting, that the test still seems to pass though oO

Comment (miri64) [reply pkg/lwip/contrib/sock/tcp/lwip_sock_tcp.c]: It didn't (probably did not clean properly on my last run). Fixed in latest PR

**Comment by benpicco** [comment]
Just squash directly, no need to bother Murdock with this twice.

**Comment by miri64** [comment]
> Just squash directly, no need to bother Murdock with this twice.

Then why activate Murdock that early in the review face ;-)?

**Comment by miri64** [comment]
Squashed.

**Comment by miri64** [comment]
Ooops, forgot a spot in `gnrc_sock`... the trouble with function pointers ;-).

**Pull request merged**